### PR TITLE
Incorporate Paris EIPs

### PR DESCRIPTION
**Pull request by alexkroeger**
This PR brings master up to date as of the Paris hard fork.

The Paris hard fork is described [here](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/paris.md) and includes the following EIPs:

- [EIP-3675: Upgrade consensus to Proof-of-Stake](https://eips.ethereum.org/EIPS/eip-3675)
- [EIP-4399: Supplant DIFFICULTY opcode with PREVRANDAO](https://eips.ethereum.org/EIPS/eip-4399)

TODO before merging:
- [x] merge https://github.com/ethereum/yellowpaper/pull/893
- [x] repo maintainers create a london branch that is a copy of master

Other TODO:

- [x] close https://github.com/ethereum/yellowpaper/pull/876
- [x] close https://github.com/ethereum/yellowpaper/pull/836

### CODE REVIEW

File: Paper.tex [hunk 1]
```diff
@@ -1422,15 +1367,11 @@ \subsection{State \& Nonce Validation}\label{sec:statenoncevalidation}
 
 \hypertarget{Phi}{}And finally we define $\Phi$, the block transition function, which maps an incomplete block $B$ to a complete block $B'$:
 \begin{eqnarray}
-\Phi(B) & \equiv & B': \quad B' = B^* \quad \text{except:} \\
-B'_{\mathrm{n}} & = & n: \quad x \leqslant \frac{2^{256}}{\hyperlink{H__d}{H_{\mathrm{d}}}} \\
-B'_{\mathrm{m}} & = & m \quad \text{with } (x, m) = \mathtt{PoW}(B^*_{\hyperlink{H_cancel_n}{\hcancel{n}}}, n, \mathbf{d}) \\
-B^* & \equiv & B \quad \text{except:} \quad {\hyperlink{Transaction_Receipt}{B^*_{\mathrm{r}}}} = \hyperlink{r}{r}(\hyperlink{Pi}{\Pi}(\Gamma(B), B))
+\Phi(B) & \equiv & B': \quad B' = B \quad \text{except:} \\
+{\hyperlink{Transaction_Receipt}{{B'_{\mathrm{H}}}_{\mathrm{r}}}} & = & \hyperlink{r}{r}(\hyperlink{Pi}{\Pi}(\Gamma(B), B))
```
Comment (alexkroeger) [inline Paper.tex:1371]: I think B_r was referring to the state root header property here, so this change would make that more clear, though I'm still a bit confused by this line.

File: README.md [hunk 2]
```diff
@@ -10,13 +10,13 @@ It is a free culture work, licensed under Creative Commons Attribution Share-Ali
 
 ## Repository Currently Outdated
```
Comment (ethosdev) [inline README.md:11]: Should this part be clarified to mention that pull requests are likely not accepted (for example https://github.com/ethereum/yellowpaper/pull/876), unless they completely specify a network upgrade?  That's why this section suggests applying for a grant?

File: Paper.tex [hunk 3]
```diff
@@ -168,12 +166,18 @@ \subsection{Which History?}
...
 
-The scheme we use in order to generate consensus is a simplified version of the GHOST protocol introduced by \cite{cryptoeprint:2013:881}. This process is described in detail in section \ref{ch:ghost}.
+Since the \textit{Paris} hard fork, reaching consensus on new blocks is managed by a protocol called the \textit{Beacon Chain}.
+It is known as the \textit{consensus layer} of Ethereum, and it defines the rules for determining the canonical history of Ethereum blocks.
+This document describes the \textit{execution layer} of Ethereum.
+The execution layer defines the rules for interacting with and updating the state of the Ethereum virtual machine.
+The consensus layer is described in greater detail in the \href{https://github.com/ethereum/consensus-specs}{consensus specifications}.
+How the consensus layer is used to determine the canonical state of Ethereum is discussed in section \ref{ch:blocktree_to_blockchain}.
 
 Sometimes, a path follows a new protocol from a particular height (block number).
```
Comment (yperbasis) [inline Paper.tex:176]: Starting from Shanghai, hard forks are triggered by block time rather than block height.

File: Paper.tex [hunk 4]
```diff
@@ -192,6 +196,7 @@ \subsection{Which History?}
 $F_{\mathrm{London}}$            & 12965000 \\
 $F_{\mathrm{Arrow Glacier}}$     & 13773000 \\
 $F_{\mathrm{Gray Glacier}}$      & 15050000 \\
+$F_{\mathrm{Paris}}$      & 15537393 \\
```
Comment (yperbasis) [inline Paper.tex:199]: I believe that 15537393 was the last PoW block, while the first PoS block was 15537394.

File: Paper.tex [hunk 5]
```diff
@@ -404,29 +409,29 @@ \subsection{The Transaction} \label{subsec:transaction}
...
 
 %\textit{TODO: Introduce logs}
 
 \begin{description}
 \item[parentHash]\linkdest{parent_Hash_H__p_def_words}{} The Keccak 256-bit hash of the parent block's header, in its entirety; formally $H_{\mathrm{p}}$.
-\item[ommersHash] The Keccak 256-bit hash of the ommers list portion of this block; formally $H_{\mathrm{o}}$.
-\item[beneficiary]\linkdest{beneficiary_H__c}{}\linkdest{H__c} The 160-bit address to which all fees collected from the successful mining of this block be transferred; formally $H_{\mathrm{c}}$.
+\item[ommersHash] A 256-bit hash field that is now deprecated due the replacement of proof of work consensus. It is now to a constant, \texttt{Keccak256(RLP([]))}; formally $H_{\mathrm{o}}$.
```
Comment (yperbasis) [inline Paper.tex:418]: The Yellow Paper uses `()` for an empty list. It's a bit confusing because it also uses `()` for an empty string (there's a comment about this in Appendix B), but cleaning that confusion up should be done in a separate PR IMO. For now it's better to use `()` for an empty list for consistency sake.

File: Paper.tex [hunk 6]
```diff
@@ -1500,24 +1416,16 @@ \subsection{Data Feeds}
...
+Providing random numbers within a deterministic system is, naturally, an impossible task. However, we can approximate with pseudo-random numbers by utilising data which is generally unknowable at the time of transacting. Such data might include the block's hash, the block's timestamp and the block's beneficiary address. In order to make it hard for malicious validators to control those values, one should use the {\small \hyperlink{blockhash}{BLOCKHASH}} operation in order to use hashes of the previous 256 blocks as pseudo-random numbers. For a series of such numbers, a trivial solution would be to add some constant amount and hashing the result.
```
Comment (yperbasis) [inline Paper.tex:1419]: Block timestamps are predictable after the Merge.

**Review by yperbasis — CHANGES_REQUESTED** [review]

File: Paper.tex (hunk 4 above)
Comment (alexkroeger) [reply Paper.tex:199]: you're correct!

File: Paper.tex [hunk 7]
```diff
@@ -404,29 +409,29 @@ \subsection{The Transaction} \label{subsec:transaction}
...
+The block in Ethereum is the collection of relevant pieces of information (known as the block \textit{header}), $H$, together with information corresponding to the comprised transactions, $\mathbf{T}$,\hypertarget{ommerheaders}{} and a now deprecated property $\mathbf{U}$ that prior to the \textit{Paris} hard fork contained headers equal to the present block's parent's parent (such blocks are known as \textit{ommers}\footnote{\textit{ommer} is a gender-neutral term to mean ``sibling of parent''; see \url{https://nonbinary.miraheze.org/wiki/Gender_neutral_language_in_English\#Aunt/Uncle}}). The block header contains several pieces of information:
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply Paper.tex:412]: good catch!

File: Paper.tex (hunk 5 above)
Comment (alexkroeger) [reply Paper.tex:418]: good catch!

File: Paper.tex (hunk 3 above)
Comment (alexkroeger) [reply Paper.tex:176]: I'll add some detail on how forks can be triggered

Comment (alexkroeger) [reply Paper.tex:176]: added!

File: Paper.tex (hunk 4 above)
Comment (alexkroeger) [reply Paper.tex:199]: fixed

File: Paper.tex (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexkroeger) [reply Paper.tex:412]: fixed

File: Paper.tex (hunk 5 above)
Comment (alexkroeger) [reply Paper.tex:418]: fixed here and at all uses in this section

File: Paper.tex (hunk 6 above)
Comment (alexkroeger) [reply Paper.tex:1419]: fixed

File: Paper.tex [hunk 8]
```diff
@@ -488,11 +499,11 @@ \subsubsection{Transaction Receipt}\linkdest{Transaction_Receipt}{}
 
 \subsubsection{Holistic Validity}
 
-\linkdest{block_validity}{}We can assert a block's validity if and only if it satisfies several conditions: it must be internally consistent with the ommer and transaction block hashes and the given transactions $B_{\mathbf{T}}$ (as specified in sec \ref{ch:finalisation}), when executed in order on the base state $\boldsymbol{\sigma}$ (derived from the final state of the parent block), result in a new state of the identity $H_{\mathrm{r}}$:
+\linkdest{block_validity}{}We can assert a block's validity if and only if it satisfies several conditions: its ommers field $B_{\mathbf{U}}$ must be an empty array and it must be internally consistent with the given transactions $B_{\mathbf{T}}$ (as specified in sec \ref{ch:finalisation}), when executed in order on the base state $\boldsymbol{\sigma}$ (derived from the final state of the parent block), result in a new state of the identity $H_{\mathrm{r}}$:
```
Comment (yperbasis) [inline Paper.tex:502]: Unclear what "it" in "it must be internally consistent with" refers to. The ommer field? Then that's wrong. Better to replace with smth. like "header transaction hash must be internally consistent with..."

File: Paper.tex [hunk 9]
```diff
@@ -488,11 +499,11 @@ \subsubsection{Transaction Receipt}\linkdest{Transaction_Receipt}{}
...
 \begin{equation}
 \begin{array}[t]{lclc}
+B_{\mathbf{U}} &\equiv& [] & \wedge \\
```
Comment (yperbasis) [inline Paper.tex:505]: Should be () instead of []

File: Paper.tex [hunk 10]
```diff
@@ -488,11 +499,11 @@ \subsubsection{Transaction Receipt}\linkdest{Transaction_Receipt}{}
...
 \linkdest{new_state_H__r}{}H_{\mathrm{r}} &\equiv& \mathtt{TRIE}(L_S(\Pi(\boldsymbol{\sigma}, B))) & \wedge \\
-\linkdest{Ommer_block_hash_H__o}{}H_{\mathrm{o}} &\equiv& \mathtt{KEC}(\mathtt{RLP}(L_H^*(B_{\mathbf{U}}))) & \wedge \\
```
Comment (yperbasis) [inline Paper.tex:495]: We still need to say that `H_o = KEC(RLP(()))`

File: Paper.tex [hunk 11]
```diff
@@ -890,7 +865,7 @@ \subsection{Execution}
 \Upsilon^{\mathrm{z}}(\boldsymbol{\sigma}, T) & \equiv & z
 \end{eqnarray}
 
-These are used to help define the \hyperlink{Transaction_Receipt}{transaction receipt} and are also used \hyperlink{sigma_n}{later} for state and nonce validation.
+These are used to help define the \hyperlink{Transaction_Receipt}{transaction receipt} and are also used \hyperlink{sigma_n}{later} for state validation.
```
Comment (yperbasis) [inline Paper.tex:868]: Here "nonce" is fine because it refers to transaction nonce.

File: Paper.tex [hunk 12]
```diff
@@ -1500,24 +1422,16 @@ \subsection{Data Feeds}
...
 
-Blockchain consolidation could be used in order to reduce the amount of blocks a client would need to download to act as a full, mining, node. A compressed archive of the trie structure at given points in time (perhaps one in every 10,000th block) could be maintained by the peer network, effectively recasting the \hyperlink{Genesis_Block}{genesis block}. This would reduce the amount to be downloaded to a single archive plus a hard maximum limit  of blocks.
+Blockchain consolidation could be used in order to reduce the amount of blocks a client would need to download to act as a full node. A compressed archive of the trie structure at given points in time (perhaps one in every 10,000th block) could be maintained by the peer network, effectively recasting the \hyperlink{Genesis_Block}{genesis block}. This would reduce the amount to be downloaded to a single archive plus a hard maximum limit  of blocks.
 
 Finally, blockchain compression could perhaps be conducted: nodes in state trie that haven't sent/received a transaction in some constant amount of blocks could be thrown out, reducing both Ether-leakage and the growth of the state database.
 
-\subsection{Scalability}
```
Comment (yperbasis) [inline Paper.tex:1513]: Why remove this section?

File: Paper.tex (hunk 8 above)
Comment (alexkroeger) [reply Paper.tex:502]: agreed, "it" just refers to the block here, i.e. the transactions it contains must be consistent w/ the state root, receipt roots. I'll make this more clear.

File: Paper.tex (hunk 9 above)
Comment (alexkroeger) [reply Paper.tex:505]: good call!

File: Paper.tex (hunk 12 above)
Comment (alexkroeger) [reply Paper.tex:1513]: I removed it because:

a) it's not essential to the spec (it didn't specify how the execution layer works, it's just a discussion of how scaling _could_ work), and 
b) it made two references to proof of work.

I think both proposals ("hierarchical structure" and "verifiable computation") are still valid under PoS, but the section also strikes me as a bit dated without discussing layer 2s.

I'm happy to keep it and add to the discussion if you think that's best, but my concern is that folks don't care what I have to say about scaling :), so I decided to just stick to the pure specification.

File: Paper.tex (hunk 9 above)
Comment (alexkroeger) [reply Paper.tex]: fixed!

File: Paper.tex (hunk 8 above)
Comment (alexkroeger) [reply Paper.tex]: I took a stab at making this more clear

File: Paper.tex (hunk 1 above)
Comment (yperbasis) [reply Paper.tex:1377]: Yes, I think that `H_r` refers to the state root. The hyperlinks look wrong here. And I think it should be `TRIE(L_S(...))` instead of `r(...)`, like in Eq (33) or (173). I suggest to change to
```
{B'_{\mathrm{H}}}_{\mathrm{r}} & = & \mathtt{TRIE}(L_{\mathrm{S}}(\hyperlink{Pi}{\Pi}(\Gamma(B), B)))
```

File: Paper.tex [hunk 13]
```diff
@@ -404,29 +415,29 @@ \subsection{The Transaction} \label{subsec:transaction}
...
-\item[mixHash]\linkdest{mixHash_H__m}{}\linkdest{H__m} A 256-bit hash which, combined with the nonce, proves that a sufficient amount of computation has been carried out on this block; formally $H_{\mathrm{m}}$.
-\item[nonce]\linkdest{block_nonce_H__n}{}\linkdest{block_nonce} A 64-bit value which, combined with the mix-hash, proves that a sufficient amount of computation has been carried out on this block; formally \hyperlink{H__n}{$H_{\mathrm{n}}$}.
+\item[prevRandao]\linkdest{prevRandao_H__a}{}\linkdest{H__a} the latest RANDAO mix\footnote{RANDAO is a pseudorandom value generated by validators on the Ethereum consensus layer. Refer to the consensus layer specs (\url{https://github.com/ethereum/consensus-specs}) for more detail on RANDAO.} of the post beacon state of the previous block; formally $H_{\mathrm{a}}$.
+\item[nonce]\linkdest{block_nonce_H__n}{}\linkdest{block_nonce} A 64-bit value that is now deprecated due to the replacment of proof of work consensus. It is set to \texttt{0x0000000000000000}; formally \hyperlink{H__n}{$H_{\mathrm{n}}$}.
```
Comment (yperbasis) [inline Paper.tex:437]: typo in "replacment"

Comment (alexkroeger) [reply Paper.tex]: fixed!

File: Paper.tex (hunk 1 above)
Comment (alexkroeger) [reply Paper.tex]: I agree, and I made the change

**Review by yperbasis — APPROVED** [review]
